### PR TITLE
Bellman Classes PR

### PR DESCRIPTION
**Pull request by Devon-FoxElster**
StaticVector, MemoFunction, ComboManager, BruteForceOptimizer, and MemoryFactory classes with initial working versions. We also added some initial testing for the next step in the process for these classes. Additionally we added .DS_Store and vscode specific settings to the gitignore,as we didn't want our personal settings contaminating the rest of the git repository nor did we want MAC-specific system settings stored on Github.

### CODE REVIEW

**Comment by Devon-FoxElster** [comment]
Changes from 07b3a88 onward are related to a Makefile issue as well as the test suites of the classes.

**Comment by Devon-FoxElster** [comment]
Commits from 0c4a867 to bbf2129 (Tests to MemoFunction all the way to BruteForceOptimizer merge) are for the final working versions of our initial classes.

File: tests/Group-10-Tests/TestComboManager.cpp [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+// Test 16: Container with duplicate elements
+TEST_CASE("ComboManager: Container with duplicate elements", "[ComboManager]") {
+    std::vector<int> numbers = {1, 2, 2, 3};
+    // With 4 items, choosing 2 gives C(4,2) = 6 index combinations,
+    // even though some value combinations may be identical.
+    cse::ComboManager<std::vector<int>> cm(numbers, 2);
+    REQUIRE(cm.TotalCombinations() == 6);
+    
+    int count = 0;
+    do {
```
Comment (LeoSSpecht) [inline tests/Group-10-Tests/TestComboManager.cpp:273]: This loop seems to be used several times on the tests. To make it a bit cleaner it could be extracted into a function or macro and used on other places

**Comment by LeoSSpecht** [comment]
# Review of `ComboManager`

## Design

### Informative identifiers
### Avoids “magic numbers”.
### Intuitive and functional interface
### Informative comments
### Consistent style
### Visual appearance

## Implementation

### No warnings during compilation
### Proper uses of namespaces
### Appropriate Modularity
### Uses modern C++ constructs
### Pointer handling
### const correctness

## Tests and error checking

### Unit testing.
### Edge cases.
### Assertions.
### Error handling

File: Group-10/Classes/ComboManager.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        k_(combination_size),
+        total_combinations_(0) {
+    if (k_ > n_) {
+      throw std::invalid_argument("Combination size cannot be greater than the number of items in the container.");
+    }
+    // Precompute the total number of combinations.
+    total_combinations_ = BinomialCoefficient_(n_, k_);
+    // Initialize indices to the first combination: [0, 1, 2, ..., k_-1].
+    indices_.resize(k_);
+    for (std::size_t i = 0; i < k_; ++i) {
```
Comment (LeoSSpecht) [inline Group-10/Classes/ComboManager.h:53]: Reset could have been called here to avoid code duplication.

**Comment by typritchett6** [comment]
### Review of Group 10 - MemoryFactory 
**Overall**
Well structured MemoryFactory class with solid testing. Some minor improvements could be made to improve performance/readability.

**Design**
•	Informative identifiers – Identifiers are named well. reservedPoint_ member variable could be named something like nextAvailable_ to better indicate what it does. GetSpace() could also be named something like GetAvailableSpace() since just GetSpace could be misleading.
•	Avoids “magic numbers” – Seems to avoid magic numbers well. 
•	Intuitive and functional interface –  The class properly includes functions like allocate and deallocate for necessary MemoryFactory functionality. Allocate could specify whether allocated objects should be deleted manually (in a comment to help avoid misuse). An extra Contains() function could be used to ensure a user only calls Deallocate() on objects from the factory. 
•	Informative comments – Comments are very informative. Some comments could be added within the Deallocate() function loop to indicate objects being reset/reinserted.
•	Consistent style – style seems to be consistent with PascalCase for functions and camelCase for member variables. 
•	Visual appearance – Code is neatly formatted with consistent indentation, etc.

**Implementation**
•	No warnings during compilation – compiling with -Werror creates an error dealing with your asserts. You need to #include <cassert> in your file I believe.
•	Proper uses of namespaces – uses namespace cse {} properly and doesn’t pollute files with extra namespaces. 
•	Adheres to the C++ standard – Looks good. Might want to look into in Deallocate() when you remove the iterator using erase then modify the targetobject after, I think erase might invalidate the iterator so targetObject might not be safe to use. 
•	Appropriate Modularity – Everything looks to be grouped together or not grouped together when necessary.
•	Uses modern C++ constructs – could consider using static constexpr int for the default allocation size which would help avoid more runtime initialization.
•	Uses standard library algorithms – could use std::find_if to look for an object in the Deallocate() function to simplify the logic.
•	Pointer handling – Could replace the allocatedBlock_ raw pointer with a smart unique pointer.
•	Const correctness – const correctness looks to be followed well.

**Tests and Error Checking**
•	Unit testing – Could add Deallocate tests for the case when an object is attempted to be deallocated but isn’t allocated by the factory (if needed). Also could add some more tests for if a bunch of objects are allocated then deallocated in some random order.
•	Edge cases – Should probably test the smallest valid allocation size (1) since that’s an edge case.
•	Assertions – asserts look to be successfully implemented to ensure valid code use (need to #include <cassert>).
•	Error handling – Can use a try catch blow to throw an exception if memory allocation fails (in AllocateObjects_() second line).

File: Group-10/Classes/MemoFunction.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  MemoFunction<int, int> memoFib([](int n) {
+    if (n <= 1) return n;
+    return recursiveFib(n - 1) + recursiveFib(n - 2);
+  });
+
+  std::cout << "Fibonacci Calculation Demo:\n";
+  std::cout << "=========================\n\n";
+
+  // Test with Fibonacci numbers
+  const int testValue = 41;
```
Comment (DominikLeisinger) [inline Group-10/Classes/MemoFunction.cpp:41]: May want to consider declaring this constant at the beginning of the function as a constexpr?

File: Group-10/Classes/MemoFunction.cpp
Comment (DominikLeisinger) [inline Group-10/Classes/MemoFunction.cpp:1]: Overall great code, very consistent and readable. Clear variable and function naming, and using std algorithms. Comments are informative for a newbie

**Review by hoggbram — COMMENTED** [review]
Design
•	Informative identifiers. Are names of variables, functions, types, etc, clear and helpful for a reader to understand the code. For example, if you need to track the number of students in a class, good variable names might be num_students or student_count, whereas you don’t want something too short to be confusing (like N or num_s) or too verbose that it’s annoying (like total_number_of_students_in_class).
Variable and function names are clear and helpful. The comments are also useful. The variable ‘w’ in Backtrack could be named more clearly, but definitely a good idea to keep it a short variable name as it’s being used in math equations later on.

•	Avoids “magic numbers”. Similar to above, you should use constant variables with informative names instead of arbitrary numbers throughout your code, and instead set constants. For example, I have a virtual CPU class with 6 virtual registers, but rather than using “6” throughout, I set static constexpr size_t NUM_REGS = 6; at the top of the class to make it clearer what the 6 refers to (and, in this case so that I can change it if needed.)
There is a constant that probably could be defined as const at the top of the header file (bestscore), or just have that be the default value when declaring the member variable.

•	Intuitive and functional interface. Does the class expose only the necessary functionality, and is it easy to use correctly while hard to misuse? Are member variables private? Is there an obvious function to use for common tasks, with relevant default arguments for parameters? Are any of the functions misleading, like potentially confusing uses of operator overloading? (Feel free to provide suggestions on how to improve the interface for better usability)
Yes it is very straightforward with its layout and there isn’t anything that could be misused.

•	Informative comments. Do comments help someone new to the code understand it more rapidly? Did the coder properly avoid redundant or obvious comments that clutter the code (e.g., ++x; // increments x). In general, comments should focus on why code is helpful, rather than belabor what that code does (though with complex code, the latter is useful too.)
Yes the comments are very clean and helpful in understanding what is going on with the code, and there are no redundant ones.

•	Consistent style. Are style choices (like naming schemes, brace placement, etc) used in a consistent and informative manner throughout the code? Is there consistency in how different naming conventions are used? (e.g., camelCase, snake_case, ALL_CAPS, or PascalCase, or prefixes and suffixes, like m_ at the beginning of a member variable.).
Yes the style is consistent.

•	Visual appearance. Is the code easy to scan through and visually well structured with good use of whitespace, indentation, and line breaks? Do there appear to be consistent limits on line widths?
My only suggestion would maybe be to put any functions that aren’t 1 line in the cpp file.

Implementation
•	No warnings during compilation. Does the code compile with the -Wall and -Wextra flags without producing warnings? One option is to add -Werror to your list of options, which turns all warnings into errors and thus halts compilation when one occurs.
When I run it, I get 3 errors: size_t hasn’t been declared and the constructor/destructor don’t have a type. It could be on my end (not having an external library installed or something) but that’s the only things I see in this part.

•	Proper uses of namespaces. Is all code within namespace cse { ... } ? Does the code avoid the generic using namespace std as well as avoiding ALL namespace pollution in header files that will spread to users of those files.
The cpp doesn’t mention the namespace, but that isn’t really applicable as there’s nothing that would need it there, so this part looks good.

•	Adheres to the C++ standard. Does the code avoid undefined behavior? If possible, make sure the code compiles and runs properly on different compilers and different platforms.
There’s a few errors on my end so I can’t check for certain, but I’m not seeing anything obvious.

•	Appropriate Modularity. Are common procedures grouped into useful helper functions? Are helper classes used where needed? Is unnecessary code duplication avoided?
Yes and I really like the use of a recursive function to find the best outcome.

•	Uses modern C++ constructs. Where applicable, does the code use range-based for loops, std::optional, or other modern C++ techniques? (In the future we will also look for proper use of templates, move semantics, lambdas, and constexpr, but only once we cover those topics).
There’s no loops or search algorithms other than the recursive function, which is just some if statements, so not applicable.

•	Uses standard library algorithms. Does the code avoid extraneous loops when equivalent functionality exists in the standard library? For example, std::fill will fill a vector, std::remove_if will remove elements from a container based on a condition, or std::is_sorted will tell you if the values in a container are in order. See https://en.cppreference.com/w/cpp/algorithm for a full list (it’s LONG, but well worth being familiar with).
Again, no loops so not applicable.

•	Pointer handling. Is memory management done carefully, with either smart pointers (std::unique_ptr, std::shared_ptr) or, only if needed, careful use of raw pointers with new and delete?
No pointers are used so not applicable.

•	const correctness. Are member functions to const if they don’t change the state of the class? Are parameters (especially reference parameters) marked const if they don’t change inside the function?
Looks correct to me.

Testing and Error Checking
•	Unit testing. Are unit tests present for all public member functions (including constructors) and well organized? Do they trigger cases that will exercise most private portions of the class’ code?
Yes all functions are tested.
•	Edge cases. Does testing cover most edge cases, as possible? For example, for pointers, does it test null pointers (if they are allowed)? Does it test for empty containers being used? Does it test for 0 or other boundary conditions on parameters?
Yes it does test edge cases, especially the cases that are rare

•	Assertions. Are assert statements used properly throughout the code to enforce invariants and ensure a valid state? (for example, checking that valid arguments are passed into functions.) Note that asserts should be used only to identify conditions that should never occur in properly working code; if code is expected to recover, error handling must be performed.
There are not any assertions in the code

•	Error handling. Are error conditions that can occur during correct execution (such as resource limitations, files not available, or division by zero) checked for and reported back with either exceptions, return conditions, or some other technique?
This kind of falls under the edge cases, as the functionality doesn’t work without this

File: Group-10/Classes/ComboManager.h
Comment (iamrb1) [inline Group-10/Classes/ComboManager.h:1]: ### Code Review for ComboManager

### Design

**Informative identifiers.** 
Names of member variables such as total_combinations_ or indices_ are straightforward and convey the meaning of functionality well. Names of functions such as NextCombo and PrevCombo convey what the function is supposed to do. Overall, great work on efficiently naming stuff to be clear to the reader for understanding the code functionality. 

**Avoids “magic numbers”.** 
No “magic numbers” are prevalent within the code (ComboManager.h is the most important to check for this standard).

**Intuitive and functional interface.** 
The class’ interface is relatively easy to use and understand with options of moving to the previous and next combination. Additionally, reset functionality has been implemented, which is nice for the user. Getters and a Binomial Coefficient method are also additionally implemented. Member variables are private and the Binomial Coefficient method is marked correctly private as the n choose k should not be accessed by the user. I did notice some confusing code related to the previous combo so any code cleanup should occur to remove the commented code or a comment should explain why the code is removed or not. 

**Informative comments.** 
Other than the comment mentioned above to be implemented, it seems that the code has useful, informative comments. I do think that lines 177 to 179 could use a comment explain for the general reader but great job with the informative comments!

**Consistent style.** 
The style is consistent throughout your code but I think that there is a discrepancy amongst your teammates. I noticed in BruteForceOptimizer that camelCase is used within their member variables whereas snake_case is utilized in yours. This should be a quick refactor though after discussing with your teammates one style to use amongst everyone.

**Visual appearance.** 
The structure is well put together and visually appealing to the reader. The code is easy to scan through. I would suggest refactoring PrevCombo to make it more visually appealing. 

### Implementation

**No warnings during compilation.** 
The Tests all compile with no warnings being produced. Well done!

**Proper uses of namespaces.** 
All code in ComboManager.h is written within the cse namespace. Code adheres to this standard well!

**Adheres to the C++ standard.** 
I believe that this code adheres to the C++ standard since there does not seem to be any use case that can enact undefined behavior whatsoever.

**Appropriate Modularity.** 
Binomial Coefficient is an example of good modularity in this code. I do think additional helped functions could be created to remove repeating code. Both NextCombo and PrevCombo could benefit from an update indices method.

**Uses modern C++ constructs.** 
Range based for loops are utilized whenever indices are not used so good implementation detail. It is possible to rewrite your algorithms to use std::begin and std::end I believe. Other than that, there is no need for modern C++ constructs learned in class so far. 

**Uses standard library algorithms.** 
There are some standard library algorithms within the link in cppreference that you can possibly use to replace some of your functionality. I do recommend utilizing std::iota for Reset method and line 53, which is essentially calling Reset yet not specified. 

**Pointer handling.** 
Pointers (raw or smart) not utilized within ComboManager so nothing to worry about for memory management. 

**const correctness.** 
Methods are marked const, for example, when applicable such as in getters. Great work!

### Testing and Error Checking

**Unit testing.** 
I believe that the test cases are well written and test all functionality written in ComboManager. It was easy to follow and was well organized.

**Edge cases.** 
Edge cases such as boundary conditions, invalid combination size, 0, and empty containers are all tested for this class. Great work!

**Assertions.** 
There are no asserts for this code. Exceptions are utilized for error handling in this class. 

**Error handling.** 
std::invalid_argument is invoked during execution properly. This is utilized for combination sizes.

**Review by KollinBartz — COMMENTED** [review]
### **_Review for BruteForceOptimizer_**

### **Design**
**Informative identifiers** - Your strings are all well worded and member functions are obvious.
**Avoids “magic numbers”** - You avoid magic numbers.
**Intuitive and functional interface** - Well done.
**Informative comments** - Your comments are really helpful and well done.
**Consistent style** - Your style seems consistent within your own files and with your teams.
**Visual appearance** - Your code is written well.

### **Implementation**
**No warnings during compilation** - I get size_t hasn’t been declared and there is no type on constructor/destructor as warnings when I run everything.
**Proper uses of namespaces** - Looks good
**Adheres to the C++ standard** - Your code does adhere to C++ standard
**Appropriate Modularity** - You have really good code here.
**Uses modern C++ constructs** - You don’t violate anything
**Uses standard library algorithms** - I don’t believe anything can really help you.
**Pointer handling** - You don’t handle them wrong
**Const correctness** - You have a good job here

### **Testing and Error Checking**
**Unit testing** - Well structured test cases.
**Edge cases** - I can’t think of any more edge cases that can show up outside of what you are testing.
**Assertions** - No asserts but you can maybe use the class assert so it still returns something testable once its implemented.
**Error handling** - You handle all errors well from what I can see.

**Review by mercere99 — APPROVED** [review]
Solid set of classes, ready to merge in.

File: tests/Group-10-Tests/TestMemoryFactory.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+// Misc. Headers
+#include <iostream>
+#include <stdexcept>
+#include <string>
+
+// Class Header
+#include "../../Group-10/Classes/MemoryFactory.h"
+
+TEST_CASE("MemoryFactory Construction", "[constructor]") {
+  SECTION("Testing default construction") {
```
[the remark this answers is not in the document]
Comment (Devon-FoxElster) [reply tests/Group-10-Tests/TestMemoryFactory.cpp:13]: Added!

File: tests/Group-10-Tests/TestMemoryFactory.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    std::string originalLiteral = "Hello There";
+    cse::MemoryFactory<std::string> stringFactory =
+        cse::MemoryFactory<std::string>(10, originalLiteral);
+    std::string* stringLiteral = stringFactory.Allocate();
+    std::string* stringLiteral2 = stringFactory.Allocate();
+    REQUIRE(*stringLiteral == originalLiteral);
+    REQUIRE(*stringLiteral2 == originalLiteral);
+    stringFactory.Deallocate(stringLiteral);
+    stringFactory.Deallocate(stringLiteral2);
+
```
[the remark this answers is not in the document]
Comment (Devon-FoxElster) [reply tests/Group-10-Tests/TestMemoryFactory.cpp:55]: Added checks for multiple Deallocation

File: Group-10/Classes/MemoryFactory.h [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+  MemoryFactory(const MemoryFactory&& other) = delete;
+  MemoryFactory& operator=(const MemoryFactory& other) = delete;
+
+  /**
+   * @brief MemoryFactory Basic Destructor
+   */
+  ~MemoryFactory() {
+    assert(reservedObjects_ == 0 &&
+           "All objects must be returned to the factory before it deletes to "
+           "avoid undefined behavior");
```
[the remark this answers is not in the document]
Comment (Devon-FoxElster) [reply Group-10/Classes/MemoryFactory.h:126]: Thanks!

File: Group-10/Classes/MemoryFactory.h [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+          "This object does not belong to this MemoryFactory");
+    }
+  }
+
+  /**
+   * @brief Returns number of remaining objects in the factory
+   *
+   * @return The number of available Objects in the MemoryFactory
+   */
+  int GetSpace() const { return allocationSize_ - reservedObjects_; }
```
[the remark this answers is not in the document]
Comment (Devon-FoxElster) [reply Group-10/Classes/MemoryFactory.h:194]: Renamed to GetAvailable

File: Group-10/Classes/MemoryFactory.h [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+      }
+    }
+  }
+
+  /**
+   * @brief Doubles the amount of allocated Objects within the MemoryFactory
+   */
+  void ExpandSpace_() {
+    AllocateObjects_(allocationSize_, defaultObject_);
+    allocationSize_ = allocationSize_ * 2;
```
[the remark this answers is not in the document]
Comment (Devon-FoxElster) [reply Group-10/Classes/MemoryFactory.h:81]: Added a constexpr to the class (value of 1.618) to calculate the expansion size

File: Group-10/Classes/BruteForceOptimizer.h [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * @brief Initiates the brute force/backtracking search to find the best subset.
+     *
+     * @return A pair containing the best total value and the corresponding subset of items.
+     */
+    std::pair<double, std::vector<Item>> FindOptimalSolution()
+    {
+        bestScore_ = -999999;
```
[the remark this answers is not in the document]
Comment (Devon-FoxElster) [reply Group-10/Classes/BruteForceOptimizer.h:71]: We made use of the numeric_limits suggestion, thank you!
